ddclient.in

```diff
@@ -2020,114 +2022,176 @@ sub geturl {
-    if ($data) {
-        $request .= "Content-Type: application/x-www-form-urlencoded\n" if $headers !~ /^Content-Type:/mi;
-        $request .= "Content-Length: " . length($data) . "\n";
-    }
-    $request .= "Connection: close\n";
-    $headers .= "\n" if $headers ne '' && substr($headers, -1) ne "\n";
-    $request .= $headers;
-    $request .= "\n";
-    # RFC 7230 says that all lines before the body must end with <cr><lf>.
-    (my $rq = $request) =~ s/(?<!\r)\n/\r\n/g;
-    $request .= $data;
-    $rq .= $data;
-
-    my %socket_args = (
-        PeerAddr => $peer,
-        PeerPort => $port,
-        Proto => 'tcp',
-        MultiHomed => 1,
-        Timeout => opt('timeout'),
-    );
-    my $socket_class = 'IO::Socket::INET';
-    if ($use_ssl) {
-        # IO::Socket::SSL will load IPv6 support if available on the system.
-        load_ssl_support;
-        $socket_class = 'IO::Socket::SSL';
-        $socket_args{SSL_ca_file} = opt('ssl_ca_file') if defined(opt('ssl_ca_file'));
-        $socket_args{SSL_ca_path} = opt('ssl_ca_dir') if defined(opt('ssl_ca_dir'));
-    } elsif ($globals{'ipv6'} || $ipversion eq '6') {
-        load_ipv6_support;
-        $socket_class = 'IO::Socket::INET6';
-    }
-    if ($ipversion eq '4') {
-        $socket_args{Domain} = PF_INET;
-        $socket_args{Family} = AF_INET;
-    } elsif ($ipversion eq '6') {
-        $socket_args{Domain} = PF_INET6;
-        $socket_args{Family} = AF_INET6;
-    } elsif ($ipversion ne '') {
-        fatal("geturl passed unsupported 'ipversion' value %s", $ipversion);
-    }
+    if (opt('curl')) {
+        if (!`curl --version 2>/dev/null`) {
+            fatal("You requested network access with curl but curl is not installed.");
+        }
+        ## Access network using curl.
+        if (!opt('exec')) {
+            debug("skipped network connection");
+            verbose("SENDING:", "%s", "${server}/${url}");
+        } elsif (($headers ne '')) {
+            my $timeout    = opt('timeout');
+            my $curlopt  = '';
+            my @head_lines = split('\n', $headers);
+            my $head_args  = '';
+            foreach my $head_line (@head_lines) {
+                $head_args .= "-H '$head_line' ";
+            }
+            $curlopt = '--ipv4' if ($ipversion eq '4');
+            $curlopt = '--ipv6' if ($ipversion eq '6');
+
+            # don't include ${url} as that might expose login credentials
+            $0 = sprintf("%s - curl (w/headers) sending to %s", $program, "https://${server}");
+            verbose("SENDING:", "curl (w/headers) to %s", "https://${server}");
+            $reply = <<`USE_CURL_WITH_HEADERS`;
+curl -s -X $method --user-agent '${program}/${version}' \\
+              --connect-timeout $timeout --max-time $timeout $curlopt \\
+              ${head_args} \\
+              --data '${data}' \\
+              --url 'https://${server}/${url}' 2>/dev/null
+USE_CURL_WITH_HEADERS
```

Comment:
Backquotes always make me nervous because it's easy to accidentally introduce a security vulnerability. Even if there is no security risk, unquoted shell metacharacters in the input variables can cause weird problems. I would be more comfortable with a subroutine that acts like backquotes except it takes a list of arguments.

Relevant SO question: https://stackoverflow.com/q/54487856.

Reply:
Learning something new every day.  I have created a `curl_cmd()` function and use that to execute curl.  I could only get it to work by passing in a single string with all the curl arguments, trying to pass in each as a separate parameter then passing the array into the `open()` function does not work... got a message back from curl that it did not recognize anything after the first argument..

In other words, this works...
```
sub curl_cmd {
    my $params = shift;
    open(my $fh, "-|", "curl " . $params);
    local $/;
    <$fh>
}
```
But this does not...
```
sub curl_cmd {
    open(my $fh, "-|", "curl" , @_);
    local $/;
    <$fh>
}
```
I suppose I could pass the arguments into `curl_cmd()` individually and then `join(' ',@_)` them all into a single string.

Thoughts?

Reply:
Further investigation... both of the options below work.  Its interesting how everything has to be separated in the second version.  Thoughts?
```
#!/usr/bin/perl -w

sub curl_cmd {
    my $params = shift;
    open(my $fh, "-|", "curl " . $params);
    local $/;
    <$fh>
}

sub curl_cmd2 {
    open(my $fh, "-|", "curl",@_);
    local $/;
    <$fh>
}

my $reply = curl_cmd("-s -X GET --url 'http://checkip.dyndns.org/'");
printf("%s\r\n", $reply);
$reply = curl_cmd2("-s","-X","GET","--url","http://checkip.dyndns.org");
printf("%s\r\n", $reply);
exit 0;
```

Reply:
Your `curl_cmd2` function is what we want as it avoids the shell. However, I suggest including the command itself in the list of arguments.

Reply:
Currently I have it implemented as `my $reply = qx{ curl --config $tmpfile 2>/dev/null };` which I think is much clearer than using file open method.  Please look at current version of the code.

The file open method is uncommon so most programmers would not immediately recognize what is going on... so unless there is a strong case for it?

Reply:
I left a separate review comment about that line, so I'm going to resolve this conversation.